[MQTT] Connect message honors LWT settings (#3006)

denisfrench opened the pull request:
The MQTT connect message was hard-coded to always be sent, with the retain flag set, regardless of UI LWT settings (#3006).

This is known to break the connection to certain MQTT brokers (Losant, AWS, possibly others).

This PR now sends the connect message to the LWT topic only if 'Send LWT to broker' is enabled, with the retain flag as per the 'Will Retain' setting.

Connection to Losant is successful using long credentials and this PR.

Fixes: #3006

denisfrench committed:
[MQTT] Connect message honors LWT settings (#3006)

TD-er commented:
Code looks good at first sight, but I wonder why there was not an active check whether it needed to send LWT at all.

denisfrench commented:
I think https://github.com/letscontrolit/ESPEasy/commit/5c9d00c0a978d4d3137b90842f092dccd7391c85 added the connect message in the days before the LWT disconnect message could be disabled.

From what I can see the connect message is just a regular message that happens to be sent to the LWT topic, tacked on to the end of the big connect function. Was probably just missed when the option to disable the proper LWT disconnect message was added, I guess.

TD-er committed:
Merge branch 'mega' into connect_msg_lwt_settings

TD-er committed:
Merge branch 'mega' into connect_msg_lwt_settings

TD-er commented on src/Controller.ino:
```diff
@@ -246,17 +246,17 @@ bool MQTTConnect(controllerIndex_t controller_idx)
+  if (MQTTclient_should_reconnect) { CPluginCall(CPlugin::Function::CPLUGIN_GOT_CONNECTED, 0); }
+  MQTTclient_should_reconnect = false;
+
+  if (ControllerSettings.mqtt_sendLWT()) { MQTTclient.publish(LWTTopic.c_str(), LWTMessageConnect.c_str(), willRetain); }
```
I think the publish of the LWT should be before the call to CPLUGIN_GOT_CONNECTED
Also if we need to send a LWT, then this publish may also fail so the connect call can return false.

TD-er replied:
Hmm on second thought, it may be better not to let this publish decide on whether it is connected or not.
The publish may fail for various reasons and there have been reports of lots of MQTT reconnect attempts on some systems.

Maybe this LWT publish state should be dealt with asynchronous, and not let the connected status depend on it.

TD-er replied:
Hmm I'm looking into the PubSubClient code and I wonder why we even try to publish to the LWT topic, as that's already been done by the connect itself.


Edit:
OK, I am for sure in need of extra coffee, the publish is sending a connected message, while we give the "disconnect" LWT message at the connect phase.
So we must publish the connected message, but I'm not 100% sure about whether we should retry sending it if unsuccessful.

denisfrench replied:
I figured successful connection cannot be determined by the connect message, given the connect message should be optional.

How critical is the "Connected" status message? It's useful for troubleshooting, but are automated processes likely to depend on it? Probably not, but that's just me.

Currently with the "Send LWT" setting disabled, a client will disconnect silently leaving the retained "Connected" LWT message, which is quite misleading.

And; GitKraken has been a revalation, thanks! Much easier to make sense of what is going on than CLI or even SourceTree :-)

TD-er replied:
I just pushed an extra commit to this PR, but based on your comment I think I still need to make a simple addition to stop trying to send the "connected" LWT message when the client is disconnected.

It is possible to flag the LWT connected message with the retain flag, so it can be used in some systems to get a quick view of who's online.
So in some systems it may be quite critical, but I don't think the return status of the connect attempt should depend on the success of the call to publish the LWT connect message.

denisfrench replied:
Fair enough; makes sense to me.

TD-er committed:
[MQTT] Process publish LWT connect message asynchronous

If publish of LWT connect message fails during connect, retry later.
The connect state should not depend on it.

TD-er committed:
[MQTT] Stop trying to send LWT connected when client disconnects

TD-er committed:
Merge branch 'mega' into connect_msg_lwt_settings

TD-er committed:
Merge branch 'mega' into connect_msg_lwt_settings

TD-er committed:
Merge branch 'mega' into connect_msg_lwt_settings

TD-er committed:
Merge branch 'mega' into connect_msg_lwt_settings
